curated_transformers/quantization/bnb/impl.py

```diff
@@ -0,0 +1,152 @@
+from typing import Callable, Optional, Set, Union
+
+import torch
+from torch import Tensor
+from torch.nn import Module, Parameter
+
+from ..._compat import bitsandbytes as bnb
+from ..._compat import has_bitsandbytes
+from ...util.pytorch import ModuleIterator, apply_to_module
+from ...util.serde import TensorToParameterConverterT
+from .config import BitsAndBytesConfig, _4BitConfig, _8BitConfig
+
+
+def prepare_for_quantization(
+    module: Module,
+    config: BitsAndBytesConfig,
+    non_quantizable_module_prefixes: Set[str],
+) -> Optional[TensorToParameterConverterT]:
+    _assert_bitsandbytes_installed()
+
+    # All inputs to quantized layers need to be of type `float16`.
+    # TODO How do we exclude non-quantizable modules from this?
```

Comment:
In the context, non-quantizable modules refers to the PyTorch `Linear` layers that we don't want to quantize. Theoretically, those layers don't need to be converted to `float16` but since if we have quantized `Linear8/4Bit` layers that come afterwards, they'll still expect `float16` inputs. I suppose the right way to do this would be to convert those inputs inside of the `Linear8/4Bit.forward()` method, but that doesn't currently happen (AFICT).

Reply:
Ah, thanks for the clarification!